### PR TITLE
gep: add GEP-3388 HTTP Retry Budget

### LINKED ISSUE #3388: Retry Budgets in HTTPRouteRetry
**What would you like to be added**:

I would like to be able to configure retry budgets in HTTPRouteRetry

**Why this is needed**:

While #1731 implements HTTPRouteRetry, we still lack the capability to set the retries using a percentage. This was part of #1731 initially but was dropped out of scope later on. 

Implementations like Linkerd rely solely on this percentage based budget to implement the retry functionality.

### PR DESCRIPTION
**Pull request by ericdbishop**
**What type of PR is this?**

/kind gep

**What this PR does / why we need it**:

To seek consensus on the ideal configuration of a "retry budget" in HTTPRoute, allowing application developers to dynamically limit the rate of client-side retries to their service based on a percentage of the active request volume.

Extends #1731

**Which issue(s) this PR fixes**:

Fixes #3388

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Comment by ericdbishop** [comment]
For the API implementation, I've compared both options of introducing retry budgets as a field of HTTPRoute

**Comment by ericdbishop** [comment]
For the API implementation, I've been comparing approaches between introducing retry budgets as a part of HTTPRoute, or implementation via policy attachments. Retry budgets are the default retry policy for Linkerd, and are highly recommended by Envoy when configuring cluster circuit breaker thresholds, so simplicity will be a priority here.

### HTTPRoute `retry` stanza

* The UX for retry budgets we'd like to match is applying the policy to a service as a whole, as opposed to individual routes. If you have multiple HTTPRoute objects pointing to the same service, you would have to configure the retry budget in multiple places.

* While Linkerd implements a retry budget alongside individual route configuration within the [ServiceProfile CRD](https://linkerd.io/2.12/reference/service-profiles/#spec), the retry budget is configured once across all routes in a single set of fields within the CRD. In practice, this works similarly to the Envoy CircuitBreaker threshold.

* If we wanted retry budgets to be configured on a per-route basis (as opposed to at the service level), it would require a change to be made in Envoy Route. I believe similar changes would need to be made for Linkerd.

### Policy Attachment

* May be more confusing from a UX perspective to configure retries in two different places (HTTPRoute for static retries, versus a policy attachment for a dynamic retry threshold).

* Ideally, we could define a single policy, targeting a service, that would dynamically configure a retry threshold based on the percentage of active requests across all routes destined for that service's backends.

* This may allow for a dynamic threshold of retries to be configured at the service level, alongside a static max number of retries on the route level. Application developers would then be allowed more granular control of which requests should be retried. For example, maybe an application developer will not want to perform retries on a specific route where requests are not idempotent, and can disable retries for that route.

---

Please let me know if my assessment of the functionality between different data plane implementations, and my understanding of how we could use policy attachments, are correct here. I would lean towards the policy attachment implementation, as it seems like the more straightforward approach, both for data plane implementations, as well as for the UX we offer to application developers. Would appreciate your thoughts here, @kflynn @mikemorris!

File: geps/gep-3388/index.md [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+### Background on implementations
+
+#### Envoy
+
+Supports configuring a [RetryBudget](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/cluster/v3/circuit_breaker.proto#envoy-v3-api-msg-config-cluster-v3-circuitbreakers-thresholds-retrybudget) CircuitBreaker threshold across a group of upstream endpoints, with the following parameters.
```
Comment (robscott) [inline geps/gep-3388/index.md:41]:
```suggestion
Supports configuring a [RetryBudget](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/cluster/v3/circuit_breaker.proto#envoy-v3-api-msg-config-cluster-v3-circuitbreakers-thresholds-retrybudget) CircuitBreaker threshold across a group of upstream endpoints, with the following parameters:
```

File: geps/gep-3388/index.md [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+* `budget_percent` Specifies the limit on concurrent retries as a percentage of the sum of active requests and active pending requests. For example, if there are 100 active requests and the budget_percent is set to 25, there may be 25 active retries. This parameter is optional. Defaults to 20%.
```
Comment (robscott) [inline geps/gep-3388/index.md:43]: To clarify, do this and the following field have a default applied if a retry budget is not actually configured in some way?

File: geps/gep-3388/index.md [hunk 3]
```diff
@@ -0,0 +1,86 @@
+# GEP-3388: HTTPRoute Retry Budget
```
Comment (robscott) [inline geps/gep-3388/index.md:1]: One of the details missing from this GEP is why this configuration should be attached at the Route level instead of the Service level. In both of the examples you link to, it appears that this policy is configured at the Service level or equivalent. It's fine if you want to cover that in a follow up, but would at least be good to make sure that's covered by an `UNRESOLVED` block to make sure we don't miss it.

File: geps/gep-3388/index.md [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+
+Configuring a limit for client retries is an important factor in building a resilient system, allowing requests to be successfully retried during periods of intermittent failure. But too many client-side retries can also exacerbate consistent failures and slow down recovery, quickly overwhelming a failing system and leading to cascading failures such as retry storms. Configuring a sane limit for max client-side retries is often challenging in complex systems. Allowing an application developer (Ana) to configure a dynamic "retry budget", reducing the risk of a high number of retries across clients, allows a service to perform as expected in both times of high & low request load, as well as both during periods of intermittent & consistent failures.
```
Comment (robscott) [inline geps/gep-3388/index.md:31]: Nit: The last sentence in this paragraph feels rather long and is a little hard to follow, it may be worth splitting it up into separate sentences.

**Review by robscott — COMMENTED** [review]
Thanks @ericdbishop!

File: geps/gep-3388/index.md (hunk 2 above)
Comment (ericdbishop) [reply geps/gep-3388/index.md:43]: Added detail to clarify. There is not default if it is not explicitly configured, the Envoy default retry circuit breaker will be a static `max_retry` value.

File: geps/gep-3388/index.md (hunk 3 above)
Comment (ericdbishop) [reply geps/gep-3388/index.md]: I added in the details from my comparison of adding retry budget via HTTPRoute versus PolicyAttachment, from [this comment](https://github.com/kubernetes-sigs/gateway-api/pull/3488#issuecomment-2548463927), into the GEP, which should help clarify where this policy should be attached. I also renamed the GEP to exclude "HTTPRoute" from the title, as I don't think it will be the right place for retry budget configuration to be added.

File: geps/gep-3388/index.md [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+
+### HTTPRoute Retry Budget
+
+* The desired UX for retry budgets is to apply the policy at the service level, rather than individually across each route targeting the service. Placing the retry budget configuration within HTTPRoute would violate this requirement, as separate HTTPRoute objects could each have routing rules targeting the same destination service, and a single HTTPRoute object can target multiple destinations. To apply a retry budget to all routes targeting a service, a user would need to duplicate the configuration across multiple routing rules.
+
+* If we wanted retry budgets to be configured on a per-route basis (as opposed to at the service level), it would require a change to be made in Envoy Route. And more than likely, similar changes would need to be made for Linkerd.
+
+## Other considerations
+
+TODO
```
Comment (mikemorris) [inline geps/gep-3388/index.md:105]:
```suggestion
* Is it worth allowing the budget to be expressed as a `Fraction` similar to `HTTPRequestMirrorFilter` as described in [GEP-3171](https://gateway-api.sigs.k8s.io/geps/gep-3171/), or is a percentage sufficient for this use case?
* As there isn't anything inherently specific to HTTP requests in either known implementation, a retry budget policy on a target Service could likely be applicable to GRPCRoute as well as HTTPRoute requests.
* While retry budgets are commonly associated with service mesh uses cases to handle many distributed clients, a retry budget policy may also be desirable for North/South implementations of Gateway API to prioritize new inbound requests and minimize tail latency during periods of service instability.
```

File: geps/gep-3388/index.md [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+
+Configuring a retry budget through a Policy Attachment may produce some confusion from a UX perspective, as users will be able to configure retries in two different places (HTTPRoute for static retries, versus a policy attachment for a dynamic retry threshold). Though this is likely a fair trade-off.
+
+Discrepancies in the semantics of retry budget behavior and configuration options between Envoy and Linkerd may require a change in either implementation to accommodate the Gateway API specification. While Envoy's `min_retry_concurrency` setting may behave similarly in practice to Linkerd's `minRetriesPerSecond`, they are not directly equivalent.
+
+A version of Linkerd's `ttl` parameter may also need to be implemented within Envoy.
```
Comment (mikemorris) [inline geps/gep-3388/index.md:79]:
```suggestion
The implementation of a version of Linkerd's `ttl` parameter within Envoy might be a path towards reconciling the behavior of these implementations, as it could allow Envoy to express a `budget_percent` and minimum number of permissible retries over a period of time rather than by tracking active and pending connections. It is not currently clear which of these models is preferable, but being able to specify a budget as requests over a window of time seems like it might offer more predictable behavior.
```

File: geps/gep-3388/index.md [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+
+* If we wanted retry budgets to be configured on a per-route basis (as opposed to at the service level), it would require a change to be made in Envoy Route. And more than likely, similar changes would need to be made for Linkerd.
+
+## Other considerations
+
+TODO
+
+## References
+
+* <https://gateway-api.sigs.k8s.io/geps/gep-1731/>
```
Comment (mikemorris) [inline geps/gep-3388/index.md:109]:
```suggestion
* <https://gateway-api.sigs.k8s.io/geps/gep-1731/>
* <https://finagle.github.io/blog/2016/02/08/retry-budgets/>
```

File: geps/gep-3388/metadata.yaml [hunk 8]
```diff
@@ -0,0 +1,35 @@
+apiVersion: internal.gateway.networking.k8s.io/v1alpha1
+kind: GEPDetails
+number: 3388
+name: HTTP Retry Budget
```
Comment (mikemorris) [inline geps/gep-3388/metadata.yaml:4]:
```suggestion
name: Retry Budgets
```

**Review by mikemorris — COMMENTED** [review]
Minor suggestions, and I think we may have found a path for reconciliation between the slightly different implementations in Envoy and Linkerd, but generally this feels like it's in great shape!

Would appreciate 👀 from @robscott again and @kflynn so hopefully we can get this merged as provisional by the deadline then move on to API design.

File: geps/gep-3388/index.md [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+
+By default, Envoy uses a static threshold for retries. But when configured, Envoy's retry budget threshold overrides any other retry circuit breaker that has been configured.
+
+#### linkerd2-proxy
+
+The Linkerd implementation of retry budgets is configured alongside service route configuration, within the [ServiceProfile CRD](https://linkerd.io/2.12/reference/service-profiles/), limiting the number of total retries for a service as a percentage of the number of recent requests. In practice, this functions similarly to Envoy's retry budget implementation, as it is configured in a single location and measures the ratio of retry requests to original requests across all traffic destined for the service.
+
+Linkerd uses [budgeted retries](https://linkerd.io/2.15/features/retries-and-timeouts/) as the default configuration to specify retries to a service, but - as of [edge-24.7.5](https://github.com/linkerd/linkerd2/releases/tag/edge-24.7.5) - supports counted retries. In all cases, retries are implemented by the `linkerd2-proxy` making the request on behalf on an application workload.
```
Comment (kflynn) [inline geps/gep-3388/index.md:52]: It's worth noting that per-route configuration in Linkerd is supported only by the various Gateway API Route resources, so it's considerably more rare for folks to use ServiceProfiles. For this reason we could consider budgeted retries less of the default configuration now.

File: geps/gep-3388/index.md [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+
+* `minRetriesPerSecond` Specifies the minimum rate of retries per second that is allowed, so that retries are not prevented when the request load is very low. The default is 10.
+
+* `ttl` A duration specifying how long requests are considered for when calculating the retry threshold. The default is 10s.
+
+### Proposed Design
+
+#### Retry Budget Policy Attachment
+
+While current retry behavior is defined at the routing rule level within HTTPRoute, exposing retry budget configuration as a policy attachment offers some advantages:
```
[the remark this answers is not in the document]
Comment (kflynn) [reply geps/gep-3388/index.md:66]: I have to respectfully disagree with @robscott here: the connection between retries and the "backend load balancer" is pretty tenuous (even in Linkerd where the component that decides which backend gets a given request is _called_ the load balancer 😉).

That does _not_ mean that I think we should have a retry policy and a circuit breaking policy and a timeout policy etc. etc., though. It means that:

a. I remain generally opposed to policy attachment for table-stakes features, and
b. If we have a catchall policy for configuring the way we interact with the backends, let's not call it BackendLBPolicy.

**Review by kflynn — COMMENTED** [review]
This is a great starting point, @ericdbishop, many thanks for diving into this! 🙂 

I left some comments and questions that, fundamentally, get into the user stories driving you to policy attachment. If you don't already know my biases here, well, we can talk about that offline (😉), but I'm asking about the stories because I want to understand what you're seeing people asking for...

File: geps/gep-3388/index.md [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+
+### HTTPRoute Retry Budget
+
+* The desired UX for retry budgets is to apply the policy at the service level, rather than individually across each route targeting the service. Placing the retry budget configuration within HTTPRoute would violate this requirement, as separate HTTPRoute objects could each have routing rules targeting the same destination service, and a single HTTPRoute object can target multiple destinations. To apply a retry budget to all routes targeting a service, a user would need to duplicate the configuration across multiple routing rules.
+
+* If we wanted retry budgets to be configured on a per-route basis (as opposed to at the service level), it would require a change to be made in Envoy Route. And more than likely, similar changes would need to be made for Linkerd.
+
+## Other considerations
+
+* Is it worth allowing the budget to be expressed as a `Fraction` similar to `HTTPRequestMirrorFilter` as described in [GEP-3171](https://gateway-api.sigs.k8s.io/geps/gep-3171/), or is a percentage sufficient for this use case? (Expressing a sub-1% budget for retries seems less necessary than for mirroring or redirecting traffic at significant scale.)
```
[the remark this answers is not in the document]
Comment (robscott) [reply geps/gep-3388/index.md:104]: +1

File: geps/gep-3388/index.md [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+
+TODO
+
+## Alternatives
+
+### HTTPRoute Retry Budget
+
+* The desired UX for retry budgets is to apply the policy at the service level, rather than individually across each route targeting the service. Placing the retry budget configuration within HTTPRoute would violate this requirement, as separate HTTPRoute objects could each have routing rules targeting the same destination service, and a single HTTPRoute object can target multiple destinations. To apply a retry budget to all routes targeting a service, a user would need to duplicate the configuration across multiple routing rules.
+
+* If we wanted retry budgets to be configured on a per-route basis (as opposed to at the service level), it would require a change to be made in Envoy Route. And more than likely, similar changes would need to be made for Linkerd.
```
[the remark this answers is not in the document]
Comment (mikemorris) [reply geps/gep-3388/index.md:100]: Agreed - I think this point was more to illustrate that per-route configuration would not be pragmatic from an implementation perspective, nor would it match the behavior currently expected by users of this functionality.

File: geps/gep-3388/index.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (mikemorris) [reply geps/gep-3388/index.md:66]: I can draft two proposed implementations (one for a new policy resource, another adding to BackendLBPolicy) in a followup PR to avoid blocking this provisional GEP on bikeshedding this now.

I do have _some_ concerns about messaging/supportability if we start glomming several discrete optional features onto a single *Policy CRD - I suppose it's not _much_ worse than what we already have with the core resources, but it's perhaps simpler for implementations to message "SpecificPolicy with X optional fields is supported" than "for BroadPolicy, X feature is supported with Y optional fields, Z feature is supported with Q optional fields, etc" and I think gets more difficult/important to promote subfields to standard channel independently rather than moving the entire resource.

[the remark this answers is not in the document]
Comment (kflynn) [reply geps/gep-3388/index.md:66]: Do you think it would help sidestep bikeshedding to describe the API simply as a stanza with relevant configuration, then have a separate discussion about where that stanza would be included?

File: geps/gep-3388/index.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (kflynn) [reply geps/gep-3388/index.md:100]: Thinking about this, and talking with some of the Linkerd maintainers, there're some... uh... _interesting_ corner cases if you do both per-Route and per-target budgeted retries, leaving me tending to support the idea of doing per-target first.

Once again, I would like to be explicit that I'm talking about supporting the _concept of per-target budgeted retries_, which is not to be conflated with yet supporting any specific manner in which one might configure such a thing. 🙂

File: geps/gep-3388/index.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (robscott) [reply geps/gep-3388/index.md:66]: My concern here is that the sheer number of resources involved in using Gateway API is overwhelming for many users (especially new ones). If we keep on with a pattern of creating a unique policy for each topic, this problem is only going to get worse. Some of the most successful Kubernetes APIs are the ones that shoved a ton of concepts into a single resource (Service, Pod, etc). Although these APIs are overloaded, they continue to be remarkably popular.

> If we have a catchall policy for configuring the way we interact with the backends, let's not call it BackendLBPolicy.

I think we'll need at least two backend policies - one for TLS config, and one for everything else. If you have any ideas for the name for the "everything else" one, I'd be open to them. I personally think `BackendLBPolicy` is ok, but can be convinced that better names exist.

Longer term, I really like the idea @ptrivedi has in https://github.com/kubernetes-sigs/gateway-api/issues/3539 that would add a new backend-focused resource to the API that could replace Service for many Gateway API users. In that proposal, it's called `EndpointSelector`, but the general idea would be to disconnect the "frontend" bits of a Service and instead have a resource exclusively focused on the backend bits. In that world, we could replace backend policies with inline fields. Not saying we should start with that for this specific GEP, but trying to provide a vision for a future that doesn't require all these backend policies.

File: geps/gep-3388/index.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (ericdbishop) [reply geps/gep-3388/index.md:100]: Just to clarify, I don't think we want per-route budgeted retries, nor have I seen anyone asking for that. I brought that up due to retry budgets initially being a proposed configuration for HTTPRetry, which would then require it to be configured at the per-route level. Some of the interest in seeing retry budgets configured per-destination in Gateway API overlaps with requests for the Envoy `retry_budget` circuit breaker threshold to be exposed in the Istio API. It's currently unavailable in Istio's Destination Rule CRD [where other Envoy circuit breaker thresholds can be configured](https://istio.io/latest/docs/reference/config/networking/destination-rule/#ConnectionPoolSettings-HTTPSettings), and will remain that way until the feature is introduced in Gateway API. I bring that up as I think it gives a better idea of what future users of Gateway API might favor as far as UX, given that Envoy and Linkerd retry budgets happen to behave similarly.

https://github.com/istio/istio/issues/27419
https://github.com/istio/api/pull/2734
https://github.com/istio/istio/pull/51823

File: geps/gep-3388/index.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (mikemorris) [reply geps/gep-3388/index.md:66]: Opened #3573 to continue an API design discussion in a followup, hoping we can get this merged as provisional as-is.

File: geps/gep-3388/index.md (hunk 9 above)
Comment (mikemorris) [reply geps/gep-3388/index.md:52]: @kflynn Any suggested language needed to clarify this before merging as provisional?

Comment (ericdbishop) [reply geps/gep-3388/index.md:52]: Addressed this [here](https://github.com/kubernetes-sigs/gateway-api/pull/3573/files#diff-a7f56eb8a3db20c8efd3ce800bce223a4fe681685f73ac077376706d8915dabaR52), just tried to capture that information along with what the UX looked like historically.

` Usage of the ServiceProfile CRD is significantly rarer today, given Linkerd's support for per-route configuration through Gateway API Route resources. Though historically, [retry budgets](https://linkerd.io/2.15/features/retries-and-timeouts/) were the default configuration to specify retries to a service, with support for counted retries later being add in [edge-24.7.5](https://github.com/linkerd/linkerd2/releases/tag/edge-24.7.5). In all cases, retries are implemented by the `linkerd2-proxy` making the request on behalf on an application workload.`

Comment (kflynn) [reply geps/gep-3388/index.md:52]: 👍 I could certainly wordsmith but this is fine, go for it! 🙂

Comment (mikemorris) [reply geps/gep-3388/index.md:52]: Updated language in followup https://github.com/kubernetes-sigs/gateway-api/pull/3573/files#diff-a7f56eb8a3db20c8efd3ce800bce223a4fe681685f73ac077376706d8915dabaR52

**Comment by ericdbishop** [comment]
@robscott @kflynn @mikemorris Seeking final approval on this so we can focus on the API design in a followup PR.

**Comment by robscott** [comment]
Thanks @ericdbishop!

/lgtm
/hold cancel
